From c2ebfad7226fd0b7caa328aa99c6dfd1def75e01 Mon Sep 17 00:00:00 2001
From: Roger Pau Monne <roger.pau@citrix.com>
Date: Thu, 19 Mar 2026 08:49:31 +0100
Subject: [PATCH] xen/mm: don't unconditionally clear _PGC_need_scrub in
 alloc_heap_pages()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

alloc_heap_pages() will unconditionally clear _PGC_need_scrub, even when
MEMF_no_scrub is requested.  This is kind of expected as otherwise some
callers will assert on seeing non-expected flags set on the count_info
field.

Introduce a new version of alloc_domheap_pages() that returns pages with
extra flags (currently only _PGC_need_scrub) set.  This new interface is
only used by populate_physmap() that knows how to deal with that flag.
Other callers are strictly kept using alloc_domheap_pages().

This fixes returning dirty pages from alloc_domheap_pages() without the
_PGC_need_scrub bit set for populate_physmap() to consume.

Fixes: 83a784a15b47 ("xen/mm: allow deferred scrub of physmap populate allocated pages")
Reported-by: Ayden Bottos <aydenbottos12@gmail.com>
Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>
diff --git a/xen/common/memory.c b/xen/common/memory.c
index b60ece8583c6..3e04e28640f1 100644
--- a/xen/common/memory.c
+++ b/xen/common/memory.c
@@ -350,7 +350,8 @@ static void populate_physmap(struct memop_args *a)
                                               &scrub_start);
 
                 if ( !page )
-                    page = alloc_domheap_pages(d, a->extent_order, memflags);
+                    page = alloc_domheap_pages_flags(d, a->extent_order,
+                                                     memflags, false);
 
                 if ( unlikely(!page) )
                 {
diff --git a/xen/common/page_alloc.c b/xen/common/page_alloc.c
index ebd88b7ce88f..dbe6486d77e6 100644
--- a/xen/common/page_alloc.c
+++ b/xen/common/page_alloc.c
@@ -923,7 +923,7 @@ static struct page_info *get_free_buddy(unsigned int zone_lo,
 static struct page_info *alloc_heap_pages(
     unsigned int zone_lo, unsigned int zone_hi,
     unsigned int order, unsigned int memflags,
-    struct domain *d)
+    struct domain *d, bool clear_scrub)
 {
     nodeid_t node;
     unsigned int i, buddy_order, zone, first_dirty;
@@ -1040,11 +1040,14 @@ static struct page_info *alloc_heap_pages(
     {
         for ( i = 0; i < (1U << order); i++ )
         {
-            if ( test_and_clear_bit(_PGC_need_scrub, &pg[i].count_info) )
+            if ( test_bit(_PGC_need_scrub, &pg[i].count_info) )
             {
                 if ( !(memflags & MEMF_no_scrub) )
                     scrub_one_page(&pg[i]);
 
+                if ( !(memflags & MEMF_no_scrub) || clear_scrub )
+                    __clear_bit(_PGC_need_scrub, &pg[i].count_info);
+
                 dirty_cnt++;
             }
             else if ( !(memflags & MEMF_no_scrub) )
@@ -2221,7 +2224,7 @@ void *alloc_xenheap_pages(unsigned int order, unsigned int memflags)
     ASSERT_ALLOC_CONTEXT();
 
     pg = alloc_heap_pages(MEMZONE_XEN, MEMZONE_XEN,
-                          order, memflags | MEMF_no_scrub, NULL);
+                          order, memflags | MEMF_no_scrub, NULL, true);
     if ( unlikely(pg == NULL) )
         return NULL;
 
@@ -2344,7 +2347,8 @@ int assign_pages(
 
         for ( i = 0; i < nr; i++ )
         {
-            ASSERT(!(pg[i].count_info & ~(PGC_extra | PGC_static)));
+            ASSERT(!(pg[i].count_info &
+                     ~(PGC_extra | PGC_static | PGC_need_scrub)));
             if ( pg[i].count_info & PGC_extra )
                 extra_pages++;
         }
@@ -2421,8 +2425,9 @@ int assign_page(struct page_info *pg, unsigned int order, struct domain *d,
     return assign_pages(pg, 1U << order, d, memflags);
 }
 
-struct page_info *alloc_domheap_pages(
-    struct domain *d, unsigned int order, unsigned int memflags)
+struct page_info *alloc_domheap_pages_flags(
+    struct domain *d, unsigned int order, unsigned int memflags,
+    bool clear_scrub)
 {
     struct page_info *pg = NULL;
     unsigned int bits = memflags >> _MEMF_bits, zone_hi = NR_ZONES - 1;
@@ -2442,12 +2447,13 @@ struct page_info *alloc_domheap_pages(
     if ( !dma_bitsize )
         memflags &= ~MEMF_no_dma;
     else if ( (dma_zone = bits_to_zone(dma_bitsize)) < zone_hi )
-        pg = alloc_heap_pages(dma_zone + 1, zone_hi, order, memflags, d);
+        pg = alloc_heap_pages(dma_zone + 1, zone_hi, order, memflags, d,
+                              clear_scrub);
 
     if ( (pg == NULL) &&
          ((memflags & MEMF_no_dma) ||
           ((pg = alloc_heap_pages(MEMZONE_XEN + 1, zone_hi, order,
-                                  memflags, d)) == NULL)) )
+                                  memflags, d, clear_scrub)) == NULL)) )
          return NULL;
 
     if ( d && !(memflags & MEMF_no_owner) )
@@ -2458,8 +2464,9 @@ struct page_info *alloc_domheap_pages(
 
             for ( i = 0; i < (1ul << order); i++ )
             {
-                ASSERT(!pg[i].count_info);
-                pg[i].count_info = PGC_extra;
+                ASSERT(!(pg[i].count_info & ~(clear_scrub ? 0
+                                                          : PGC_need_scrub)));
+                pg[i].count_info |= PGC_extra;
             }
         }
         if ( assign_page(pg, order, d, memflags) )
diff --git a/xen/include/xen/mm.h b/xen/include/xen/mm.h
index 48931d57c46b..3a45126cbbbc 100644
--- a/xen/include/xen/mm.h
+++ b/xen/include/xen/mm.h
@@ -130,8 +130,21 @@ void get_outstanding_claims(uint64_t *free_pages, uint64_t *outstanding_pages);
 
 /* Domain suballocator. These functions are *not* interrupt-safe.*/
 void init_domheap_pages(paddr_t ps, paddr_t pe);
-struct page_info *alloc_domheap_pages(
-    struct domain *d, unsigned int order, unsigned int memflags);
+
+/*
+ * Direct callers of alloc_domheap_pages_flags() must sanitize the contents
+ * of the count_info field to ensure no unexpected flags are leaked.
+ */
+struct page_info *alloc_domheap_pages_flags(
+    struct domain *d, unsigned int order, unsigned int memflags,
+    bool clear_scrub);
+
+static inline struct page_info *alloc_domheap_pages(
+    struct domain *d, unsigned int order, unsigned int memflags)
+{
+    return alloc_domheap_pages_flags(d, order, memflags, true);
+}
+
 void free_domheap_pages(struct page_info *pg, unsigned int order);
 unsigned long avail_domheap_pages_region(
     unsigned int node, unsigned int min_width, unsigned int max_width);
